synapse/dendrite/jobs.py

```diff
@@ -0,0 +1,214 @@
+
+    def complete(self, job):
+        '''
+        Complete the job provided.
+
+        The job provided is transitioned to a state of 'completed'.
+
+        Args:
+            job (dict):
+                The job to be completed.
+        '''
+        tufo = self.core.getTufoByIden(job.get('iden'))
+        if tufo:
+            logger.debug('Completing job %s', tufo)
+            self.core.setTufoProps(tufo, status='completed', completed_at=self._currTimestamp())
+
+    def fail(self, job):
+        '''
+        Fail the job provided.
+
+        The job provided is transitioned back to a state of 'queued', this will make the job available again for
+        retrieval from the queue.
+
+        Args:
+            job (dict):
+                The job to fail.
+        '''
+        tufo = self.core.getTufoByIden(job.get('iden'))
+        if tufo:
+            # for now, just log and requeue the job
+            logger.debug('Failing job %s', tufo)
+            self.core.setTufoProps(tufo, status='failed', failed_at=self._currTimestamp())
+
+    def qsize(self, queue, status='queued'):
+        '''
+        Determine the size of the queue specified.
+
+        By default, the number of 'queued' jobs is returned. A status is optionally specified to determine the number
+        of jobs in other states, e.g. 'working' or 'completed'.
+
+        Args:
+            queue (str):
+                The name of the queue to determine the size of.
+
+            status (str):
+                The job status to predicate on while determining the size.
+
+        Returns:
+            int: Size of the queue.
+        '''
+        return len(self._jobsByQueue(queue, status))
+
+    def isEmpty(self, queue, status='queued'):
+        '''
+        Determine whether the queue specified is empty.
+
+        A convenience method for determining whether a queue's size is 0.
+
+        Args:
+            queue (str):
+                The name of the queue to determine the size of.
+
+            status (str):
+                The job status to predicate on while determining the size.
+
+        Returns:
+            bool: True if the queue is empty, False otherwise.
+        '''
+        return self.qsize(queue, status) == 0
+
+    def stats(self, status='any'):
+        '''
+        Returns a histogram showing job count by queue name.
+
+        The default status argument is 'any', which is a pseudo status for collecting all jobs regardless of status.
+
+        Args:
+            status (str):
+                The status to filter the jobs by, e.g. 'queued', 'working', 'completed'.
+
+        Returns:
+            collections.defaultdict: A dictionary containing a histogram of count by queue name.
+        '''
+        if status == 'any':
+            return self.core.getStatByProp('histo', 'dendrite:job:queue')
+
+        histo = collections.defaultdict(int)
+        for row in self.core.getJoinByProp('dendrite:job:status', status):
```

Comment:
While this is functionally correct, the only reason to go directly to the row level APIs is if it allows you some optimization over lifting tufos.  In this case, it may actually be slowing you down since access to dendrite:job:queue properties would be a dictionary lookup rather than O(N) iteration through all the properties of the dendrite:job.  Love the idea of stats plumbing though!